### ISSUE TITLE
Faulty d2g in ObjectNav baselines

### ISSUE THREAD

**Issue by joel99** [allenai/allenact#274]
Hello, I was taking a look at the pretrained DDPPO ObjectNav agent (in `objectnav_baselines`) and running it through ai2thor-docker + allenact. I was curious what the existing failure modes were. I found two surprising issues from evaluation output:

1. The agent is occasionally successful even when D2G is > 1.0m. When I plot the D2G recorded in metrics (`dist_to_target`), I see some successes even when the agent appears far away from goal.

![image](https://user-images.githubusercontent.com/14226466/115469346-7a466480-a202-11eb-8f46-095950f618f0.png)

It's hard to reproduce this precisely but e.g. I had it occur on the following scenes:
![image](https://user-images.githubusercontent.com/14226466/115469515-c8f3fe80-a202-11eb-9dfd-a919de8ced76.png)
where `optimal_d2g` is the D2G reported at the first timestep.

2. `path_to_target` appears to be invalid.
![image](https://user-images.githubusercontent.com/14226466/115469774-40c22900-a203-11eb-9eaf-1a57874c8f5b.png)
This doesn't always happen, but in investigating the above I saw this peculiar `path_to_target` in the task_info for `FloorPlan_Val2_4_Vase_6`. These don't seem to be valid movements for an agent. Does `path_to_target` mean something else?

These results come from a full, headless docker setup. The dockerfile is as follows:
```
FROM ai2thor-docker:latest

COPY requirements.txt dev_requirements.txt /app/
COPY allenact_plugins/robothor_plugin/extra_requirements.txt /app/
RUN pip install -r requirements.txt; pip install -r dev_requirements.txt; pip install -r extra_requirements.txt
...
```
- ai2thor-docker is from current master.

cc: @Lucaweihs , who put together the baselines. Thank you for the wonderful resource!

**Comment by jordis-ai2** [comment]
Hi Joel. I think that things might still be working as expected, if I understood the issue.

For Q1, it is possible that the agent lies at an Euclidean distance < 1 m  from the target while the geodesic distance remains larger (there could be an object crossing the line between the agent and the target). Without more details it is difficult to confirm there is a bug, so it would be great if you could also output the agent location and/or the Euclidean distance to the target.

For Q2, the provided optimal path assumes an agent with continuous motion, so the distance covered by that path is a lower bound.

**Comment by Lucaweihs** [comment]
Hi @joel99 ,

I was just responding and now see that @jordis-ai2 beat me to the punch.

1. I will dive into this tomorrow to see if I can can reproduce what you're seeing, as Jordi said it would be helpful to have any additional information if possible. If you're running inference using AllenAct can you give us the `metrics_....json` that was generated?
2. As noted by Jordi, the `path_to_target` is computed using a ["navmesh"](https://docs.unity3d.com/Manual/nav-BuildingNavMesh.html) within Unity which assumes that the agent can move any distance in any direction (but not through walls/objects of course). This is very efficient but is means that the path cannot always be actually achieved by the agent.

**Comment by joel99** [comment]
Hello all,

re: 1, I attached a subset (Final D2G > 1m and Success == True) of the metrics output, hopefully this helps, and a snippet which should reproduce those orange dots. Note it's been renamed to a txt to make github happy, it should be changed back to json.
re: 2, I see, that makes sense. 

```
import numpy as np
import seaborn as sns
import json
import pandas as pd

path = 'metrics__abbrev.json'
with open(path, 'r') as f:
    info = json.load(f)
tasks = info['tasks']

def extract_ep(ep):
    task_info = ep['task_info']
    def _to_pos(d):
        return np.array([d['x'], d['y'], d['z']])
    def _to_path(path):
        return np.stack([_to_pos(p)[[0, 2]] for p in path], 0)
    def _dist(a, b):
        return np.sqrt(((a - b)**2).sum())
    start = _to_pos(task_info['followed_path'][-1])
    end = _to_pos(task_info['followed_path'][0])
    return {
        'd2g': ep['dist_to_target'],
        'env_d2g': ep.get('env_d2g', None),
        'steps': ep['ep_length'],
        'path': _to_path(task_info['followed_path']),
        'reward': ep['reward'],
        'spl': ep['spl'],
        'success': ep['success'],
        'scene': task_info['scene'],
        'id': task_info['id'],
        'mirrored': task_info['mirrored'],
        'optimal_steps': len(task_info['path_to_target']),
        'optimal_path': task_info['path_to_target'],
        'optimal_d2g': task_info['distance_to_target'],
        'obj_cat': task_info['object_type'],
        'start': start,
        'end': end,
        'dist': _dist(start, end),
        'actions': task_info['taken_actions']
    }

extracted = [extract_ep(ep) for ep in tasks]
df = pd.DataFrame(extracted)
ax = sns.scatterplot(x='optimal_d2g', y='d2g', hue='success', data=df)

```
[metrics__abbrev.txt](https://github.com/allenai/allenact/files/6353212/metrics__abbrev.txt)

**Comment by jordis-ai2** [comment]
I noticed distance cache inconsistencies when we were sampling episodes from more than one scene in a task sampler. Testing with the fix should produce something like
![image](https://user-images.githubusercontent.com/39779720/115806186-228f3100-a39b-11eb-9ad0-50a9ef1acc8f.png)

**Comment by Lucaweihs** [comment]
Thanks for bring this to our attention @joel99, as Jordi mentioned above: this is fixed in #275 which is now merged.
